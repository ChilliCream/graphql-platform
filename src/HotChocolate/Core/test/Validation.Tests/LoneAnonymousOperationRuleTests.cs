using Microsoft.Extensions.DependencyInjection;
using Xunit;

namespace HotChocolate.Validation
{
    public class LoneAnonymousOperationRuleTests
        : DocumentValidatorVisitorTestBase
    {
        public LoneAnonymousOperationRuleTests()
            : base(builder => builder.AddOperationRules())
        {
        }

        [Fact]
        public void QueryContainsOneAnonymousOperation()
        {
            ExpectValid(@"
                {
                    dog {
                        name
                    }
                }
            ");
        }

        [Fact]
        public void QueryWithOneAnonymousAndOneNamedOperation()
        {
            ExpectErrors(@"
                {
                    dog {
                        name
                    }
                }

                query getName {
                    dog {
                        owner {
                            name
                        }
                    }
                }
            ",
            t => Assert.Equal(
                "GraphQL allows a short‐hand form for defining query " +
                "operations when only that one operation exists in the " +
                "document.", t.Message));
        }

        [Fact]
        public void QueryWithTwoAnonymousOperations()
        {
            ExpectErrors(@"
                {
                    dog {
                        name
                    }
                }

                {
                    dog {
                        name
                    }
                }
            ",
            t => Assert.Equal(
                "GraphQL allows a short‐hand form for defining query " +
                "operations when only that one operation exists in the " +
                "document.", t.Message));
        }

        [Fact]
        public void MultipleNamedOperations()
        {
            ExpectValid(@"
                query Foo { 
                    dog {
                        name
                    }
                }
                query Bar { 
                    dog {
                        name
                    }
                }
            ");
        }

        [Fact]
        public void AnonymousOperationWithFragment()
        {
            ExpectValid(@"
                {
                    ...Foo
                }
                fragment Foo on Query { 
                    dog {
                        name
                    }
                }
            ");
        }

        [Fact]
        public void AnonymoutOperationWithAMutation()
        {
            ExpectErrors(@"
                {
                    dog {
                        name
                    }
                } 
                mutation Foo {
                    fieldB
                }
            ");
        }

        [Fact]
        public void AnonymoutOperationWithASubscription()
        {
            ExpectErrors(@"
                {
                    dog {
                        name
                    }
                } 
                subscription Foo {
                    newMessage
                }
            ");
        }
    }
}
