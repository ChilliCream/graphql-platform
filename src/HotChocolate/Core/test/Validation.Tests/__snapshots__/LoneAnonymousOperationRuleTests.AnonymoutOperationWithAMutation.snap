[
  {
    "Message": "GraphQL allows a short‐hand form for defining query operations when only that one operation exists in the document.",
    "Code": null,
    "Path": null,
    "Locations": [
      {
        "Line": 2,
        "Column": 17
      }
    ],
    "Extensions": {
      "operations": 2,
      "specifiedBy": "http://spec.graphql.org/June2018/#sec-Lone-Anonymous-Operation"
    },
    "Exception": null,
    "SyntaxNode": {
      "Kind": "OperationDefinition",
      "Location": {
        "Start": 17,
        "End": 139,
        "Line": 2,
        "Column": 17
      },
      "Name": null,
      "Operation": "Query",
      "VariableDefinitions": [],
      "Directives": [],
      "SelectionSet": {
        "Kind": "SelectionSet",
        "Location": {
          "Start": 17,
          "End": 139,
          "Line": 2,
          "Column": 17
        },
        "Selections": [
          {
            "Kind": "Field",
            "Alias": null,
            "Arguments": [],
            "SelectionSet": {
              "Kind": "SelectionSet",
              "Location": {
                "Start": 43,
                "End": 113,
                "Line": 3,
                "Column": 25
              },
              "Selections": [
                {
                  "Kind": "Field",
                  "Alias": null,
                  "Arguments": [],
                  "SelectionSet": null,
                  "Location": {
                    "Start": 69,
                    "End": 95,
                    "Line": 4,
                    "Column": 25
                  },
                  "Name": {
                    "Kind": "Name",
                    "Location": {
                      "Start": 69,
                      "End": 95,
                      "Line": 4,
                      "Column": 25
                    },
                    "Value": "name"
                  },
                  "Directives": []
                }
              ]
            },
            "Location": {
              "Start": 39,
              "End": 113,
              "Line": 3,
              "Column": 21
            },
            "Name": {
              "Kind": "Name",
              "Location": {
                "Start": 39,
                "End": 44,
                "Line": 3,
                "Column": 21
              },
              "Value": "dog"
            },
            "Directives": []
          }
        ]
      }
    }
  }
]
